src/ProgressOnderwijsUtils/Collections/Tree.cs

```diff
@@ -132,37 +183,22 @@ public static Tree<T> BuildRecursively<T>(T root, IReadOnlyDictionary<T, IReadOn
         /// i.e. it will be called for the root node last, and that root node may differ from the initial root node as subtrees have already been pruned.
         /// </summary>
         [Pure]
-        public static Tree<T>? Where<T>(this Tree<T> tree, Func<Tree<T>, bool> retainSubTree)
-        {
-            var reconstruct = new Stack<Tree<T>>(16);
-            var todo = new Stack<Tree<T>>(16);
-            todo.Push(tree);
-            while (todo.Count > 0) {
-                var next = todo.Pop();
-                reconstruct.Push(next);
-                var children = next.Children;
-                for (var i = children.Count - 1; i >= 0; i--) {
-                    todo.Push(children[i]);
-                }
-            }
-            var tmp = new List<Tree<T>>();
-            while (reconstruct.Count > 0) {
-                var next = reconstruct.Pop();
-                var kidCount = next.Children.Count;
-                for (var i = 0; i < kidCount; i++) {
-                    var maybeKid = todo.Pop();
-                    if (retainSubTree(maybeKid)) {
-                        tmp.Add(maybeKid);
-                    }
-                }
-                todo.Push(Node(next.NodeValue, tmp.ToArray()));
-                tmp.Clear();
-            }
-            return todo.Pop() is var finalTree && retainSubTree(finalTree) ? finalTree : null;
-        }
+        public static Tree<T>? Where<TTree, T>(this IRecursiveStructure<TTree, T> tree, Func<TTree, bool> retainSubTree)
+            where TTree : IRecursiveStructure<TTree, T>
+            => CachedTreeBuilder<TTree, T>.Resolve(tree.TypedThis, o => o.Children.Where(retainSubTree), (o, kids) => Node(o.NodeValue, kids));

```

Comment:
Ahh, wat is deze where makkerlijker.

Een verschil: deze where is top-down ipv. bottom up.

Reply:
Heb jij wel eens gezien hoe een echte boom eruit ziet?

Reply:
jaja ;-).